File: programmers/캐시.cpp
#include <iostream>
#include <vector>

using namespace std;

string string_to_lower(string s){
    string result = "";
    for (int i=0;i<s.size();i++){
        char now = s[i];
        if ('A'<=now && now<='Z'){
            now = now + 'a'-'A';
            
        } 
        result += now;
    }
    return result;
}

int solution(int cacheSize, vector<string> cities) {
    // 최근에 들어온 데이터가 뒤에
    vector<string> cache;
    int answer = 0;

    if (cacheSize == 0) {
        return 5*cities.size();
    }
    
    for (int i=0;i<cities.size();i++){
        string now = string_to_lower(cities[i]);
        bool isCached = false;
        for (int j=0;j<cache.size();j++){
            if (now == cache[j]){
                isCached = true;
                cache.erase(cache.begin()+j);
                cache.push_back(now);
                break;
            }
        }

        if (isCached){
            answer+=1;
        } else {
            if (cacheSize == cache.size()){
                cache.erase(cache.begin());
            }
            cache.push_back(now);
            answer+=5;
        }
    }
    return answer;
}

int main(){
    vector<string> cities;
    cities.push_back("Jeju");
    cities.push_back("Pangyo");
    cities.push_back("Seoul");
    cities.push_back("NewYork");
    cities.push_back("LA");
    cities.push_back("Jeju");
    cities.push_back("Pangyo");
    cities.push_back("Seoul");
    cities.push_back("NewYork");
    cities.push_back("LA");
    cout << solution(3, cities) << "\n";

}

File: programmers/level2/캐시.py
def solution(cacheSize, cities):

    if cacheSize == 0:
        return len(cities) * 5
    cache = []
    answer = 0

    cities = [city.lower() for city in cities]

    for city in cities:
        is_is_cache = city in cache
        if is_is_cache:
            # 캐시에 있는거 빼서 맨뒤에 추가
            if len(cache) == cacheSize:
                cache.remove(city)
            cache.append(city)
            answer += 1
        else:
            # 캐시에서 제일 오래 된 거 빼고 이거 추가
            if len(cache) == cacheSize:
                cache.pop(0)
            cache.append(city)
            answer += 5

    return answer
